src/gc.c

```diff
@@ -500,142 +500,134 @@ STATIC_INLINE void gc_queue_big_marked(jl_ptls_t ptls, bigval_t *hdr,
     ptls->gc_cache.nbig_obj = nobj + 1;
 }
 
-// `gc_setmark_big` and `gc_setmark_pool` can be called concurrently on
-// multiple threads. In all cases (except gc-debug),
-// the functions atomically sets the mark bits and updates the metadata
-// if the bits are changed.
+// `gc_setmark_tag` can be called concurrently on multiple threads.
+// In all cases, the functions atomically sets the mark bits and returns
+// the GC bits set as well as if the tag was unchanged by this thread.
 // All concurrent calls on the same object guarantees to be setting the
 // bits to the same value.
 // For normal objects, this is the bits with only `GC_MARKED` changed to `1`
 // For buffers, this is the bits of the owner object.
 // For `mark_reset_age`, this is `GC_MARKED` with `GC_OLD` cleared.
-static inline uint16_t gc_setmark_big(jl_ptls_t ptls, jl_taggedvalue_t *o,
-                                      int8_t mark_mode, uintptr_t tag)
+// The return value is `1` if the object was not marked before and the object
+// needs to be scanned. Returning `0` can happen if another thread
+// marked it in parallel or if the object is known to not reference
+// any object.
```

Comment:
A reference always means a pointer in this file so there's no reference to itself (unless the object has a field pointing to itself).

This comment is actually a little out of date so I will fix (remove it)